src/Orleans/Core/ClusterClient.cs

```diff
@@ -118,26 +126,31 @@ public void Abort()
 
         private async Task Stop(bool gracefully)
         {
-            if (this.state == LifecycleState.Disposed) return;
-
+            if (this.IsDisposing) return;
             using (await this.initLock.LockAsync().ConfigureAwait(false))
             {
                 if (this.state == LifecycleState.Disposed) return;
-                if (gracefully)
+                try
                 {
-                    Utils.SafeExecute(() => this.runtimeClient.Disconnect());
+                    this.state = LifecycleState.Disposing;
+                    if (gracefully)
+                    {
+                        Utils.SafeExecute(() => this.runtimeClient.Disconnect());
+                    }
+
+                    Utils.SafeExecute(() => this.runtimeClient.Reset(gracefully));
+                    this.Dispose(true);
+                }
+                finally
+                {
+                    // If disposal failed, the system is in an invalid state.
+                    if (this.state == LifecycleState.Disposing) this.state = LifecycleState.Invalid;
                 }
-
-                Utils.SafeExecute(() => this.runtimeClient.Reset(gracefully));
-                this.Dispose();
             }
         }
 
         /// <inheritdoc />
-        public void Dispose()
-        {
-            this.Dispose(true);
-        }
+        void IDisposable.Dispose() => this.Abort();
```

Comment:
small breaking change here. Why was this turned into an explicit implementation of IDisposable? I think it's negligible, since the concrete class is internal, so the explicit implementation might not break anybody, but not sure I understand the reasoning